Adapt MSUnmerged to sync RSE data with MongoDB properly

todor-ivanov opened the pull request:
Fixes #10797 

#### Status
ready

#### Description
The current PR is supposed to resolve few  issues which were introduced with adding MongoDB as a backend database for the service, which are making the service not backward compatible:
 * After preserving the timestamps into the database we are getting into the situation that a RSE would be retried only once in a through out a span of a single Rucio Consistency Monitor poling cycle. now we should be able to retry the RSE in case it was not previously completely cleaned
 * If we have a fresh new record for the RSE in the Rucio Consistency Monitor and we do not wipe out all the provious file records for that RSE in MongoDB, then we are about to start accumulating old records. With this PR we are resetting the object if a brand new record is present in Rucio Consistency Monitor.

#### Is it backward compatible (if not, which system it affects?)
NO

#### Related PRs
No additional configurations PRs are required, but it is worth mentioning the previous PR:
https://github.com/dmwm/WMCore/pull/10895

#### External dependencies / deployment changes
<Does it require deployment changes? Does it rely on third-party libraries?>

todor-ivanov committed:
Adapt MSUnmerged to sync RSE data with MongoDB properly

Add resetRSE method && resetRSE before starting a fresh cleanup && Implement readRSEFromMongoDB with projections.

ResetRSE only if a new record in RucioConMon is present.

Fix some typos and missing logs

fixes from reviews

Fix bug with field type transformation during write to DB for fields of gen/filter type.

Try to recover from NonMasterError during write operation to MongoDB.

Move retries of MongoDB write operation in the RSE writeRSEToMongoDB method.

Fix retry logic condition and log messages.

Fix missing argument for updateRSETimestamps.

amaltaro commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -431,15 +431,32 @@ def consRecordAge(self, rse):
-            msg += "Skipping it in the current run."
-            self.logger.info(msg)
-            raise MSUnmergedPlineExit(msg)
+            if rse['isClean']:
```
With the current definition of `isClean`, doesn't it mean that the service will keep trying to delete files until it manages to delete every single file that is listed as deletable? I'm afraid we need to redefine how we consider an RSE as clean or not.

Another option, which is likely better, would be to add a new key/value pair to the MSUnmergedRSE object to hold the original timestamp - when the file dump was collected - provided by Rucio ConMon. Such that we end up with an expression like:
```
isConsNewer = self.rseConsStats[rseName]['end_time'] > self.rseTimestamps[rseName]['dumpTStamp']
```

todor-ivanov replied:
With the current definition of `isClean` we refresh/consider as new the RSE cleanup status on every RucioConMon polling cycle. Meaning if RucioConMon provides a new dump of the RSE, we reset everything - counters and cleanup status (keeping only the timestamps/updating them as usual), and we retry a fresh cleanup regarding the new RSE situation. So if there have been some leftovers from previous Runs those must have been detected by RucioConMon and will get into the our next iteration. If we have cleaned the RSE last time. We should consider it unclean now and retry again - in case there is nothing to be deleted it will enter immediately in clean status, if there is something new - it will be deleted together with leftovers (if any).

The idea of preserving the RucioConMon timestamp is not bad, let me check what I can do to put it there.

todor-ivanov replied:
Hi @amaltaro, I did implement this and as simple as it sounds from the line you suggested, at the end it became a quite horrible change, because I also had to check the status of the RucioConMon record for the RSE not only here  but also in the `updateRSETimestamps` method, so we can be sure to update the record for this timestamp in MongoDB only if the RucioConMon has done a complete and successful scan of the RSE, basically re implementing the whole `consRecordAge` method, and at the end we will get exactly the same result. I think this is going to make things more error prone than simplifying them.

amaltaro commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -431,15 +431,32 @@ def consRecordAge(self, rse):
-            self.logger.info(msg)
-            raise MSUnmergedPlineExit(msg)
+            if rse['isClean']:
+                msg += "And the RSE has been cleaned during the last Rucio Consistency Monitor pooling cycle."
```
`pooling` --> `polling`. Same for a few lines below.

todor-ivanov replied:
thanks ! :)

vkuznet replied:
It is always dangerous to use constructs like `if rse['isClean']` since you never know rse data type at run-time, it is much safer to use the following `if rse and 'isClean' in rse`. Here you always check that rse exists, and then check if certain key exists in a directionary.

todor-ivanov replied:
thanks @vkuznet ,  I was brave here because we make sure on several other places that we do have a common, well defined and universal `rse` data structure, which should always include that field. But you are correct, staying on the safe side and following good practice principles, this should be written the way you suggest here. I will fix that.

amaltaro commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -74,13 +82,17 @@ def buildMongoProjection(self, fullRSEToDB=False):
         return mongoProjection
 
-    def readRSEFromMongoDB(self, collection):
+    def readRSEFromMongoDB(self, collection, useProjection=False):
```
Don't forget updating the docstring please.

todor-ivanov replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -145,3 +157,22 @@ def writeRSEToMongoDB(self, collection, fullRSEToDB=False, overwrite=False):
+            resetDoc['counters'] = self['counters']
+
+        self.update(resetDoc)
+        writeResult = self.writeRSEToMongoDB(collection, fullRSEToDB=True, overwrite=True)
```
Isn't `fullRSEToDB` meant to be parameterized? It's not clear to me why it's set to True here.

todor-ivanov replied:
And I have just found a bug which caused the the transformation of the data type for some fields when we writing/resetting the RSE record to MongoDB, e.g. 
```
In [67]: rse.resetRSE(msUnmergedColl)
Out[67]: True

In [68]: rse
Out[68]: 
{'name': 'T2_DE_DESY',
 'pfnPrefix': None,
 'isClean': False,
 'timestamps': {'rseConsStatTime': 0.0,
  'prevStartTime': 0.0,
  'startTime': 0.0,
  'prevEndTime': 0.0,
  'endTime': 0.0},
 'counters': {'totalNumFiles': 0,
  'dirsToDeleteAll': 0,
  'dirsToDelete': 0,
  'filesToDelete': 0,
  'deletedSuccess': 0,
  'deletedFail': 0},
 'files': {'allUnmerged': [],
  'toDelete': {},
  'protected': {},
  'deletedSuccess': [],
  'deletedFail': []},
 'dirs': {'allUnmerged': set(), 'toDelete': set(), 'protected': set()},
 '_id': ObjectId('61a0e052b713a7692b8475f9')}

In [69]: msUnmergedColl.find_one(rse.mongoFilter)
Out[69]: 
{'_id': ObjectId('61a0e052b713a7692b8475f9'),
 'name': 'T2_DE_DESY',
 'pfnPrefix': None,
 'isClean': False,
 'timestamps': {'rseConsStatTime': 0.0,
  'prevStartTime': 0.0,
  'startTime': 0.0,
  'prevEndTime': 0.0,
  'endTime': 0.0},
 'counters': {'totalNumFiles': 0,
  'dirsToDeleteAll': 0,
  'dirsToDelete': 0,
  'filesToDelete': 0,
  'deletedSuccess': 0,
  'deletedFail': 0},
 'files': {'allUnmerged': [],
  'toDelete': [],
  'protected': [],
  'deletedSuccess': [],
  'deletedFail': []},
 'dirs': {'allUnmerged': [], 'toDelete': [], 'protected': []}}
```

Notice the type of all the fields which are of filter/generator type (e.g. `rse['files']['toDelete']`). The fix is in place already, but this one was tricky to debug and I was just lucky to notice it before merging the PR.

amaltaro reviewed (changes requested):
Todor, I had a quick review and left some comments for your consideration.

In addition to that, I wonder if this is actually a required change for the upcoming CMSWEB release?

todor-ivanov commented:
Thanks for your reviews @amaltaro and @vkuznet .  I've addressed what I feel safe enough to do at this very moment. I have  tested extensively the logic involved. Also found a bug and fixed that one too. I will try to merge this tonight because I'd like to push this with the tag I am about to cut any moment now. Please feel free to leave comments if you happen to look at it again after I have merged it.

todor-ivanov committed:
Pylint fixes
